core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/CompactExpressionEvaluator.scala

```diff
@@ -0,0 +1,382 @@
+      // To determine which entities to fetch from the database
+      val rootEntityType = rootEntityTypeOpt.get
+
+      val entityLookups: Seq[ExpressionLookup] = expressionEvaluationContext.expression match {
+        case None             => Seq.empty
+        case Some(expression) => parseLookups(expression)
+      }
+
+      // Parse all input expressions and collect their lookups and parsed trees
+      val inputExpressionData = gatherInputsResult.processableInputs.toSeq.map { input =>
+        val terraExpressionParser = AntlrTerraExpressionParser.getParser(input.expression)
+        val visitor = new CompactEvaluateVisitor()
+        val parsedTree = terraExpressionParser.root()
+        val inputLookups: Seq[ExpressionLookup] = visitor.visit(parsedTree)
+        (input, parsedTree, inputLookups)
+      }
+
+      // Gather all ExpressionLookups from all inputs
+      val allLookups = inputExpressionData.flatMap(_._3)
+
+      // Build query plans for all lookups combined
+      val queryPlans = if (entityType != rootEntityType && entityLookups.nonEmpty) {
+        // Use the entityLookups to determine the starting point for the relationChains in the query plans
+        val entityRelationChain = entityLookups.flatMap(_.relations.map(_.attributeName().getText)).toList
+        val baseQueryPlans = buildQueryPlans(allLookups)
+
+        baseQueryPlans.map { plan =>
+          plan.copy(relationChain = entityRelationChain ++ plan.relationChain)
+        }
+      } else {
+        buildQueryPlans(allLookups)
+      }
+
+      // Execute all query plans
+      val queryFutures: Seq[Future[Seq[ExpressionAndResult]]] = queryPlans.map { plan =>
+        executeQueryPlan(workspaceId, entityType, entityName, rootEntityType, plan, entityLookups)
+      }
+
+      Future.sequence(queryFutures).flatMap { allQueryResults =>
+        val combinedExpressionAndResults: Seq[ExpressionAndResult] = allQueryResults.flatten
+
+        // Determine the correct root entity names based on rootEntityType
+        val rootEntityNames = if (entityType == rootEntityType) {
+          // Entity type matches root entity type, use the original entity name
+          Seq(entityName)
+        } else {
+          // Entity type differs from root entity type, extract entity names from query results
+          // The query results contain entities of the rootEntityType
+          combinedExpressionAndResults.flatMap(_._2.keys).distinct
+        }
+
+        // Process each input separately using the combined query results
+        val inputFutures: Seq[Future[Seq[(ExpressionEvaluationSupport.EntityName, SubmissionValidationValue)]]] =
+          inputExpressionData.map { case (input, parsedTree, inputLookups) =>
+
+            // Filter ExpressionAndResults relevant to this input
+            val relevantResults = combinedExpressionAndResults.filter { case (expression, _) =>
+              inputLookups.exists(_.expression == expression)
+            }
```

Comment:
What are the scales of the `inputExpressionData` and `combinedExpressionAndResults` Seqs here? I am wondering about the nested loop and if it would be better to restructure this data into something with more random access instead of using `.filter`.